opalog/parser_test.go

```diff
@@ -91,48 +114,345 @@ func TestCompositesWithRefs(t *testing.T) {
+	assertParseOneExpr(t, "composites", "[{14.2: true, \"a\": null}] != {foo: {a.b[0]: [10]}}", NewBuiltinExpr(VarTerm("!="), left2, right2))
 }
 
-func assertTermNotEqual(t *testing.T, x *Term, y *Term) {
-	if x.Equal(y) {
-		t.Errorf("Failure on non-equality: \n%s and \n%s\n", x, y)
-	}
+func TestMiscBuiltinExpr(t *testing.T) {
+	xyz := VarTerm("xyz")
+	assertParseOneExpr(t, "empty", "xyz()", NewBuiltinExpr(xyz))
+	assertParseOneExpr(t, "single", "xyz(abc)", NewBuiltinExpr(xyz, VarTerm("abc")))
+	assertParseOneExpr(t, "multiple", "xyz(abc, {\"one\": [1,2,3]})", NewBuiltinExpr(xyz, VarTerm("abc"), ObjectTerm(Item(StringTerm("one"), ArrayTerm(NumberTerm(1), NumberTerm(2), NumberTerm(3))))))
+}
+
+func TestNegatedExpr(t *testing.T) {
+	assertParseOneTermNegated(t, "scalars 1", "not true", BooleanTerm(true))
+	assertParseOneTermNegated(t, "scalars 2", "not \"hello\"", StringTerm("hello"))
+	assertParseOneTermNegated(t, "scalars 3", "not 100", NumberTerm(100))
+	assertParseOneTermNegated(t, "scalars 4", "not null", NullTerm())
+	assertParseOneTermNegated(t, "var", "not x", VarTerm("x"))
+	assertParseOneTermNegated(t, "ref", "not x[y].z", RefTerm(VarTerm("x"), VarTerm("y"), StringTerm("z")))
+	assertParseOneExprNegated(t, "vars", "not x = y", NewBuiltinExpr(VarTerm("="), VarTerm("x"), VarTerm("y")))
+
+	ref1 := RefTerm(VarTerm("x"), VarTerm("y"), StringTerm("z"), VarTerm("a"))
+
+	assertParseOneExprNegated(t, "membership", "not x[y].z[a] = \"b\"", NewBuiltinExpr(VarTerm("="), ref1, StringTerm("b")))
+	assertParseOneExprNegated(t, "misc. builtin", "not sorted(x[y].z[a])", NewBuiltinExpr(VarTerm("sorted"), ref1))
+}
+
+func TestPackage(t *testing.T) {
+	ref1 := RefTerm(DefaultRootDocument, StringTerm("foo"))
+	assertParsePackage(t, "single", "package foo", &Package{Path: ref1.Value.(Ref)})
+	ref2 := RefTerm(DefaultRootDocument, StringTerm("f00"), StringTerm("bar_baz"), StringTerm("qux"))
+	assertParsePackage(t, "multiple", "package f00.bar_baz.qux", &Package{Path: ref2.Value.(Ref)})
+	assertParseError(t, "non-ground ref", "package foo[x]")
+}
+
+func TestImport(t *testing.T) {
+	assertParseImport(t, "single", "import foo", &Import{Path: VarTerm("foo").Value})
+	ref := RefTerm(VarTerm("foo"), StringTerm("bar"), StringTerm("baz")).Value
+	assertParseImport(t, "multiple", "import foo.bar.baz", &Import{Path: ref})
+	assertParseImport(t, "single alias", "import foo as bar", &Import{Path: VarTerm("foo").Value, Alias: Var("bar")})
+	assertParseImport(t, "multiple alias", "import foo.bar.baz as qux", &Import{Path: ref, Alias: Var("qux")})
+	assertParseError(t, "non-ground ref", "import foo[x]")
+}
+
+func TestRule(t *testing.T) {
+
+	assertParseRule(t, "identity", "p = true :- true", &Rule{
+		Name:  Var("p"),
+		Value: BooleanTerm(true),
+		Body: []*Expr{
+			&Expr{Terms: []*Term{BooleanTerm(true)}},
+		},
+	})
+
+	assertParseRule(t, "identity 2", "p :- true", &Rule{
```

Comment:
What does this rule mean?  Same as the one above?  Is this just a weird case that just falls out of the implementation?